frame/assets/src/lib.rs

```diff
@@ -1438,149 +1429,505 @@ impl<T: Config> Pallet<T> {
-				WithdrawConsequence::ReducedToZero(rest)
+				if keep_alive {
+					(WithdrawConsequence::WouldDie, None)
+				} else {
+					(WithdrawConsequence::ReducedToZero(rest), maybe_new_frozen)
+				}
 			} else {
 				// NOTE: this assumes (correctly) that the token won't be a provider. If that ever
 				// changes, this will need to change.
-				WithdrawConsequence::Success
+				(WithdrawConsequence::Success, maybe_new_frozen)
 			}
 		} else {
-			WithdrawConsequence::NoFunds
+			(WithdrawConsequence::NoFunds, None)
 		}
 	}
 
-	fn increase_balance(
+	// Maximum `amount` that can be passed into `can_withdraw` to result in a WithdrawConsequence
+	// of Success.
+	fn decreasable_balance(
 		id: T::AssetId,
-		beneficiary: T::AccountId,
+		who: &T::AccountId,
+		keep_alive: bool,
+		respect_frozen: RespectFrozen,
+	) -> Result<T::Balance, Error<T>> {
+		let details = match Asset::<T>::get(id) {
+			Some(details) => details,
+			None => return Err(Error::<T>::Unknown),
+		};
+		ensure!(!details.is_frozen, Error::<T>::Frozen);
+
+		let account = Account::<T>::get(id, who);
+		ensure!(!account.is_frozen, Error::<T>::Frozen);
+
+		let amount = match (keep_alive, respect_frozen, T::Freezer::frozen_balance(id, who)) {
+			(_, Respect, Some(frozen)) => {
+				// Frozen balance that we respect: account CANNOT be deleted
+				let required = frozen.checked_add(&details.min_balance).ok_or(Error::<T>::Overflow)?;
+				account.balance.saturating_sub(required)
+			}
+			(true, _, _) => {
+				account.balance.saturating_sub(details.min_balance)
+			}
+			(_, _, _maybe_frozen) => {
+				// No frozen balance or not respecting it: account can be deleted. If f.is_some(),
+				// a notification via `melted` is required.
+
+				// NOTE: this assumes (correctly) that the token won't be a provider. If that ever
+				// changes, this will need to change.
+
+				account.balance
+			}
+		};
+		Ok(amount.min(details.supply))
+	}
+
+	/// Make preparatory checks for debiting some funds from an account. Flags indicate requirements
+	/// of the debit.
+	///
+	/// - `amount`: The amount desired to be debited. The actual amount returned for debit may be
+	///   less (in the case of `best_effort` being `true`) or greater by up to the minimum balance
+	///   less one.
+	/// - `keep_alive`: Require that `target` must stay alive.
+	/// - `respect_frozen`: Respect any freezes on the account or token (or not).
+	/// - `best_effort`: The debit amount may be less than `amount`.
+	///
+	/// On success, the amount which should be debited (this will always be at least `amount` unless
+	/// `best_effort` is `true`) together with an optional value indicating the argument which must
+	/// be passed into the `melted` function of the `T::Freezer` if `Some`.
+	///
+	/// If no valid debit can be made then return an `Err`.
+	fn prep_debit(
+		id: T::AssetId,
+		target: &T::AccountId,
+		amount: T::Balance,
+		keep_alive: bool,
+		respect_frozen: RespectFrozen,
+		best_effort: bool,
```

Comment:
These 3 inputs are always together and seem to all represent some kind of setting on how you want to handle an asset.

Maybe it would make sense to turn them into a struct.